security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -267,6 +287,38 @@ func (s *sdsservice) FetchSecrets(ctx context.Context, discReq *xdsapi.Discovery
 	return sdsDiscoveryResponse(secret, discReq.Node.Id)
 }
 
+func (s *sdsservice) Stop() {
+	s.closing <- true
+}
+
+func (s *sdsservice) clearStaledClientsJob() {
+	s.ticker = time.NewTicker(s.tickerInterval)
+	for {
+		select {
+		case <-s.ticker.C:
+			s.clearStaledClients()
+		case <-s.closing:
+			if s.ticker != nil {
+				s.ticker.Stop()
+			}
+		}
+	}
+}
+
+func (s *sdsservice) clearStaledClients() {
+	log.Debug("start staled connection cleanup job")
+	sdsClientsMutex.Lock()
+	defer sdsClientsMutex.Unlock()
+
+	len := len(staledClientKeys)
+	for i := 0; i < len; i++ {
+		log.Debugf("remove staled clients %+v", staledClientKeys[i])
+		delete(sdsClients, staledClientKeys[i])
+	}
+
+	staledClientKeys = staledClientKeys[len:]
```

Comment:
it isn't immediately obvious what this is doing to me

Reply:
add comments